client/fingerprint/consul.go

```diff
@@ -298,6 +304,72 @@ func (cfs *consulFingerprintState) grpcTLSPort(info agentconsul.Self) (string, b
 	return fmt.Sprintf("%d", int(p)), ok
 }
 
+func (cfs *consulFingerprintState) dnsPort(info agentconsul.Self) (string, bool) {
+	p, ok := info["DebugConfig"]["DNSPort"].(float64)
+	return fmt.Sprintf("%d", int(p)), ok
+}
+
+// dnsAddr fingerprints the Consul DNS address, but only if Nomad can use it
+// usefully to provide an iptables rule to a task
+func (cfs *consulFingerprintState) dnsAddr(logger hclog.Logger) func(info agentconsul.Self) (string, bool) {
+	return func(info agentconsul.Self) (string, bool) {
+
+		var listenOnEveryIP bool
+
+		dnsAddrs, ok := info["DebugConfig"]["DNSAddrs"].([]string)
+		if ok {
+			for _, dnsAddr := range dnsAddrs {
+				dnsAddr = strings.TrimPrefix(dnsAddr, "tcp://")
+				dnsAddr = strings.TrimPrefix(dnsAddr, "udp://")
+
+				parsed, err := netip.ParseAddrPort(dnsAddr)
+				if err != nil || !parsed.IsValid() {
+					logger.Warn("could not parse Consul addresses.dns config", "value", dnsAddr)
+					return "", false // response is somehow malformed
+				}
+
+				// only addresses we can use for an iptables rule from a
+				// container to the host will be fingerprinted
+				if parsed.Addr().IsUnspecified() {
+					listenOnEveryIP = true
+					break
+				}
+				if !parsed.Addr().IsLoopback() {
+					return parsed.Addr().String(), true
+				}
+			}
+		}
+
+		// if Consul DNS is bound on 0.0.0.0, we want to fingerprint the private
+		// IP (or at worst, the public IP) of the host so that we have a valid
+		// IP address for the iptables rule
+		if listenOnEveryIP {
+
+			privateIP, err := sockaddr.GetPrivateIP()
+			if err != nil {
+				logger.Warn("could not query network interfaces", "error", err)
+				return "", false // something is very wrong, so bail out
+			}
+			if privateIP != "" {
+				return privateIP, true
+			}
+			publicIP, err := sockaddr.GetPublicIP()
```

Comment:
I imagine the "fallback to private or public IP" test hits privateIP above so misses this path, leaving it untested, but to do anything about that (like a mockable interface) seems more trouble than its worth.  I just wanted to call it out in case you can think of a simple solution off the top of your head.

Reply:
Yeah, that'd have been nice, but it'd mostly be testing the library code anyways, which hopefully is well-tested already.

Reply:
Ran thru with a `-cover` profile and this is now the only bit we're missing coverage on.

Reply:
> it'd mostly be testing the library code anyways

I do try to avoid that, yeah.  the missing coverage here is not how sockaddr does its work, but that we do ask it to do its work, when appropriate.  the logic here is so simple though that it seems fine to skimp on that little bit of coverage.